### PR TITLE
kernel/semaphore: Manage the list of semaphore holder

### PR DESCRIPTION
**Pull request by jeongarmy**
1. A list of all semaphores in kernel region
- Add semaphore to the list when it is created(initialized)
- Remove semaphore from the list when it is destroyed

2. A list of all holders for each semaphore
- Save tcb of holder when it takes semaphore
- Clear its own holder data when it gives semaphore
- Post semaphore and Clear its own holder data when task is terminated with taking semaphore

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [1a2739974d0df7a899ac557568360039047bc69f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1a2739974d0df7a899ac557568360039047bc69f] - Code Rule Check Result:
lib/libc/semaphore/sem_init.c:80: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/include/semaphore.h:129: ERROR: [IDT_M_TAB] please, no space before tabs
os/kernel/semaphore/sem_destroy.c:100: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/kernel/semaphore/sem_recover.c:104: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

File: os/include/tinyara/semaphore.h [hunk 1]
```diff
@@ -92,6 +92,9 @@ struct nsem_inode_s {
 };
 #endif
 
+#define MAX_PID_MASK           (CONFIG_MAX_TASKS - 1)
```
Comment (jeongchanKim) [inline os/include/tinyara/semaphore.h]: We already have a definition for PIDHASH.
Let's use this define.

File: lib/libc/semaphore/sem_init.c [hunk 2]
```diff
@@ -120,6 +143,12 @@ int sem_init(FAR sem_t *sem, int pshared, unsigned int value)
 		sem->holder.htcb = NULL;
 		sem->holder.counts = 0;
 #endif
+#endif
+
+#if defined(CONFIG_APP_BINARY_SEPARATION) && defined(__KERNEL__)
+		/* Register semaphore in kernel region for kernel resourse management */
```
Comment (jeongchanKim) [inline lib/libc/semaphore/sem_init.c]: resourse -> resource

**Comment by seinfra** [comment]
Target : [57251b932dc1488dc41d27ea020c693c0767ee91] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [57251b932dc1488dc41d27ea020c693c0767ee91] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [ace4cd443586e30a7c3223c083a751b887c080ee] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [ace4cd443586e30a7c3223c083a751b887c080ee] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [eb010e88efb0711d60a9df97b88b7cafbc0213e2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [eb010e88efb0711d60a9df97b88b7cafbc0213e2] - Code Rule Check OK.

File: lib/libc/semaphore/sem_init.c [hunk 3]
```diff
@@ -60,11 +60,34 @@
...
 #include <tinyara/debug/sysdbg.h>
 #endif
 
+#if defined(CONFIG_APP_BINARY_SEPARATION) && defined(__KERNEL__)
+sq_queue_t g_sem_list;
+
+static void sem_register(FAR sem_t *sem)
+{
+	sem_t *sem_ptr;
+
```
Comment (sijupunnnoose) [inline lib/libc/semaphore/sem_init.c]: Should we update list in irqsave? I cant think of a scenario now..

File: os/kernel/semaphore/sem_destroy.c [hunk 4]
```diff
@@ -85,6 +88,26 @@
...
  ****************************************************************************/
+#ifdef CONFIG_APP_BINARY_SEPARATION
+extern sq_queue_t g_sem_list;
+
+static void sem_unregister(FAR sem_t *sem)
+{
+	sem_t *sem_ptr;
+	irqstate_t flags;
+
+	sem_ptr = (sem_t *)sq_peek(&g_sem_list);
```
Comment (sijupunnnoose) [inline os/kernel/semaphore/sem_destroy.c]: Should we update list in irqsave? I cant think of a scenario now..

File: os/kernel/semaphore/sem_recover.c [hunk 5]
```diff
@@ -88,6 +92,27 @@
...
+extern sq_queue_t g_sem_list;
+
+static void sem_tcb_recover(FAR struct tcb_s *tcb)
+{
+	sem_t *sem;
+	irqstate_t flags;
+
+	sem = (sem_t *)sq_peek(&g_sem_list);
+	while (sem) {
+		if (sem->holders[PIDHASH(tcb->pid)] == tcb) {
```
Comment (sijupunnnoose) [inline os/kernel/semaphore/sem_recover.c]: Move irqsave  out of "if" statement.

File: os/include/semaphore.h [hunk 6]
```diff
@@ -106,18 +109,28 @@ struct semholder_s {
...
+#endif
 	int16_t semcount;			/* >0 -> Num counts available */
 	/* <0 -> Num tasks waiting for semaphore */
 	/* If priority inheritance is enabled, then we have to keep track of which
 	 * tasks hold references to the semaphore.
 	 */
 
 	uint8_t flags;			/* See definitions for the struct sem_s flags */
+#ifdef CONFIG_APP_BINARY_SEPARATION
+	FAR struct tcb_s *holders[CONFIG_MAX_TASKS];	/* List of holders */
```
Comment (sangam-swami) [inline os/include/semaphore.h]: do we need to initialize the holders value?

File: os/kernel/semaphore/sem_destroy.c (hunk 4 above)
Comment (jeongarmy) [reply os/kernel/semaphore/sem_destroy.c]: I think it needs because of global variable, g_sem_list

**Comment by jeongarmy** [comment]
@sijupunnnoose sem_reset just resets a semaphore to a specific value. I think I don't need to do something  because this is not related holder information.

File: os/include/semaphore.h [hunk 7]
```diff
@@ -106,18 +109,28 @@ struct semholder_s {
 #endif
 #endif							/* CONFIG_PRIORITY_INHERITANCE */
 
+#ifdef CONFIG_APP_BINARY_SEPARATION
+#define IS_KERNEL_SEM(sem) ((sem->flags & FLAGS_KERNELSEM) != 0)
```
Comment (Taejun-Kwon) [inline os/include/semaphore.h]: One question, is named semaphore also included in kernel sem?

Comment (jeongarmy) [reply os/include/semaphore.h]: Right, it is created in kernel space.

File: os/kernel/semaphore/sem_destroy.c (hunk 4 above)
Comment (sijupunnnoose) [reply os/kernel/semaphore/sem_destroy.c]: Should we have peek also inside irqsave is was my thought is.

File: lib/libc/semaphore/sem_init.c (hunk 3 above)
Comment (sijupunnnoose) [reply lib/libc/semaphore/sem_init.c]: I think we have to add entire block inside irqsave

File: os/kernel/semaphore/sem_destroy.c [hunk 8]
```diff
@@ -85,6 +88,26 @@
...
+	while (sem_ptr) {
+		if (sem_ptr == sem) {
+			flags = irqsave();
+			sq_rem((FAR sq_entry_t *)sem, &g_sem_list);
```
Comment (yashwanth686007) [inline os/kernel/semaphore/sem_destroy.c]: I think no need to traverse through entire list. That logic is taken care in sq_rem.

File: os/include/semaphore.h (hunk 6 above)
Comment (yashwanth686007) [inline os/include/semaphore.h]: The holders queue can also be dynamically created instead of static.

**Comment by seinfra** [comment]
Target : [bab63720815ac7c5c6eb80c0c3d3fee8c32e0a09] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [bab63720815ac7c5c6eb80c0c3d3fee8c32e0a09] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [844e7695df22a6cdccb762774092d88556c8f085] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [844e7695df22a6cdccb762774092d88556c8f085] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fde06a9bebd220e3209abb0d0c5e05b9e5e1aec5] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fde06a9bebd220e3209abb0d0c5e05b9e5e1aec5] - Code Rule Check OK.

**Pull request closed without merging**